libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -51,21 +50,42 @@ bool KnowledgeBase::knownToBeDifferent(YulString _a, YulString _b)
 	return false;
 }
 
+optional<u256> KnowledgeBase::differenceIsKnownConstant(YulString _a, YulString _b)
+{
+	// Try to use the simplification rules together with the
+	// current values to turn `sub(_a, _b)` into a constant.
+
+	Expression expr1 = simplify(FunctionCall{{}, {{}, "sub"_yulstring}, util::make_vector<Expression>(Identifier{{}, _a}, Identifier{{}, _b})});
+	if (Literal const* value = get_if<Literal>(&expr1))
+		return valueOfLiteral(*value);
+
+	return {};
+}
+
 bool KnowledgeBase::knownToBeDifferentByAtLeast32(YulString _a, YulString _b)
 {
 	// Try to use the simplification rules together with the
 	// current values to turn `sub(_a, _b)` into a constant whose absolute value is at least 32.
 
-	Expression expr1 = simplify(FunctionCall{{}, {{}, "sub"_yulstring}, util::make_vector<Expression>(Identifier{{}, _a}, Identifier{{}, _b})});
-	if (holds_alternative<Literal>(expr1))
-	{
-		u256 val = valueOfLiteral(std::get<Literal>(expr1));
-		return val >= 32 && val <= u256(0) - 32;
-	}
+	if (optional<u256> difference = differenceIsKnownConstant(_a, _b))
+		return difference >= 32 && difference <= u256(0) - 32;
 
 	return false;
 }
 
+bool KnowledgeBase::knownToBeZero(YulString _a)
+{
+	return knownConstant(_a) == u256{};
+}
+
+optional<u256> KnowledgeBase::knownConstant(YulString _a)
```

Comment:
```suggestion
optional<u256> KnowledgeBase::valueIfKnownConstant(YulString _a)
```
although, I'll admit that that's not particularly catchy...

Reply:
And thinking about it here the name technically also works (i.e. returing a constant, if known, rather than returning whether it's known that it's constant), but it still does sound like a boolean...

Reply:
In general, the KnowledgeBase changes would make a quick PR with an easy merge if extracted.